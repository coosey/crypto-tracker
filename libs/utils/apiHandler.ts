import { NextApiRequest, NextApiResponse } from 'next';

interface ApiHandlerConfig {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  headers?: Record<string, string>;
  cacheConfig?: {
    enabled?: boolean;
    sMaxAge?: number;
    staleWhileRevalidate?: number;
  };
  queryDefaults?: Record<string, string | number>;
  requiredParams?: string[];
};

function urlQueryBuilder(query: Record<string, any>, urlBuilder: string) {
  if (Object.keys(query).length === 0) return urlBuilder;
  const queryParams = new URLSearchParams();
  for (const [key, value] of Object.entries(query)) {
    if (Array.isArray(value)) {
      value.forEach(v => queryParams.append(key, v));
    } else if (value) {
      queryParams.append(key, value as string);
    }
  }
  return `${urlBuilder}?${queryParams.toString()}`;
};

export function createApiHandler<T>(config: ApiHandlerConfig) {
  return async function handler(
    req: NextApiRequest,
    res: NextApiResponse<T | { error: string }>
  ) {
    try {
      if (config?.cacheConfig?.enabled !== false) {
        res.setHeader(
          'Cache-Control',
          `public, s-maxage=${config?.cacheConfig?.sMaxAge ?? 1800}, stale-while-revalidate=${config?.cacheConfig?.staleWhileRevalidate ?? 3600}`
        );
      }

      // Handle query parameters with defaults
      const query = {
        ...config?.queryDefaults,
        ...req?.query,
      };

      // Validate required parameters
      if (config?.requiredParams) {
        const missingParams = config?.requiredParams?.filter?.(
          param => !query?.[param]
        );
        if (missingParams?.length > 0) {
          return res.status(400).json({
            error: `Missing required parameters: ${missingParams?.join(', ')}`,
          });
        }
      }
      // Create a copy of the request query to avoid mutation
      let queryCopy = { ...req.query };

      // Build final URL
      let finalUrl = config?.url;
      // Replace path parameters if they exist in the URL i.e. /coins/{id}/market_chart
      if (config?.url?.includes?.('{')) {
        for (const [key, value] of Object.entries(queryCopy)) {
          if (config?.url?.includes?.(`{${key}}`)) {
            finalUrl = finalUrl?.replace?.(`{${key}}`, value as string);
            delete queryCopy[key];
          }
        }
      }
      // Then append query parameters
      if (Object.keys(queryCopy)?.length > 0) {
        finalUrl = urlQueryBuilder(queryCopy, finalUrl);
      }

      const response = await fetch(finalUrl, {
        method: config?.method || 'GET',
        headers: {
          'content-type': 'application/json',
          ...config?.headers,
        },
      });

      if (!response?.ok) {
        throw {
          status: response?.status,
          message: `API request failed with status ${response?.status}`,
        };
      }
      let data = await response.json();

      res.status(200).json(data);
    } catch (error) {
      console.error('API handler error:', error);
      res.status((error as any)?.status || 500).json({
        error: (error as Error)?.message || 'An unknown error occurred',
      });
    }
  };
}