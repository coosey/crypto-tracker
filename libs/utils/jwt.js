import jwt from 'jsonwebtoken';

export const generateToken = (payload, secret, expiresIn = '1h') => {
  return jwt.sign(payload, secret, { expiresIn });
};
