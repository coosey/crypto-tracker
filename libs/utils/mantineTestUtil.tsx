import { MantineProvider } from '@mantine/core';
import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

const MantineTestProvider = ({ children }: { children: React.ReactNode }) => {
  return <MantineProvider>{children}</MantineProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: MantineTestProvider, ...options });

export * from '@testing-library/react';
export { customRender as render };