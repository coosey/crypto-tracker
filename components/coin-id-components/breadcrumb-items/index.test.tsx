
import { BreadCrumbItems } from '.';
import { render, screen } from '@libs/utils/mantineTestUtil';
describe('BreadCrumbItems Component', () => {
  it('Should render BreadcrumbItems Component correctly', () => {
    render(
      <BreadCrumbItems
        name="Home"
      />
    );
    const breadcrumbComponent = screen.getByText('Home Price');
    expect(breadcrumbComponent).toMatchSnapshot();
  });
  
});