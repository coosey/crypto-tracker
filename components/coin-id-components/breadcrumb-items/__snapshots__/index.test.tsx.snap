// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadCrumbItems Component Should render BreadcrumbItems Component correctly 1`] = `
<span
  class="m_f678d540 mantine-Breadcrumbs-breadcrumb breadcrumb_nolink"
>
  Home Price
</span>
`;
